tests/test_metrics.py

```diff
@@ -0,0 +1,215 @@
+    SUBSCRIPTIONS_IN_PROGRESS,
+    update_subscription_metrics,
+)
+
+from .conftest import enum_get_query, longout_get_query, nan_get_query
+
+
+@pytest.fixture(autouse=True)
+def clear_metrics():
+    """Reset all metrics before and after each metrics-based test"""
+    REGISTRY.clear()
+    # Must forcibly reload the modules in order to a) recreate the metrics and
+    # b) re-create the "metrics_middleware" wrapped function that is used when
+    # creating the application.
+    importlib.reload(coniql.metrics)
+    importlib.reload(coniql.app)
+    yield
+    REGISTRY.clear()
+
+
+@pytest.mark.asyncio
+async def test_metrics_endpoint(client: TestClient):
+    """Test the metrics endpoint starts and returns basic info"""
+
+    resp = await client.get("/metrics")
+    assert resp.status == 200
+
+    text = await resp.text()
+    expected_collectors = REGISTRY.collectors
+
+    # Check the endpoint returns the same number of collectors as the registry
+    assert text.count("HELP") == len(expected_collectors)
+    assert text.count("TYPE") == len(expected_collectors)
+
+    # And check that each named expected item exists
+    assert ACTIVE_CHANNELS.name in text
+    assert DROPPED_UPDATES.name in text
+    assert REQUEST_EXCEPTIONS.name in text
+    assert REQUEST_TIME.name in text
+    assert REQUESTS.name in text
+    assert SUBSCRIPTIONS_IN_PROGRESS.name in text
+
+
+@pytest.mark.asyncio
+async def test_metrics_query_count(ioc, client: TestClient):
+    """Test metrics for counting GraphQL Queries and Mutations"""
+
+    resp = await client.get("/ws", params={"query": longout_get_query})
+    assert resp.status == 200
+
+    resp = await client.get("/metrics")
+    assert resp.status == 200
+
+    text = await resp.text()
+
+    # The request total metrics
+    assert 'coniql_request_total{path="/ws",route="middleware"} 1' in text
+    assert 'coniql_request_total{route="GraphQL"} 1' in text
+
+
+@pytest.mark.asyncio
+async def test_metrics_processing_time(ioc, client: TestClient):
+    """Test metrics for duration of requests.
+    Note must make two requests as the @timer annotation won't finish timing a
+    request until after the request has finished, but the data returned is
+    created in the middle of the request so is not included"""
+
+    resp = await client.get("/ws", params={"query": longout_get_query})
+    assert resp.status == 200
+
+    resp = await client.get("/metrics")
+    assert resp.status == 200
+
+    text = await resp.text()
+
+    # Only expect 1 as the current request isn't tracked
+    assert "coniql_request_processing_seconds_count 1" in text
+    assert 'coniql_request_processing_seconds{quantile="0.5"}' in text
+    assert 'coniql_request_processing_seconds{quantile="0.9"}' in text
+    assert 'coniql_request_processing_seconds{quantile="0.99"}' in text
+    assert "coniql_request_processing_seconds_sum" in text
+
+
+@pytest.mark.asyncio
+async def test_metrics_active_channels(ioc, client: TestClient):
```

Comment:
Do you want to add an xfail test to show the problem with counting active channels you mention in the description?

Reply:
I've implemented the necessary introspection in aioca, https://github.com/dls-controls/aioca/pull/38, but it isn't released yet - I'll see about doing that

Reply:
Implemented in https://github.com/dls-controls/coniql/pull/53/commits/e9e08e032151a90537479a4a5dcad284a5d23d41